PNG driver reworked to use libpng 1.6 simplified API

ibrahn opened the pull request:
Wrapped libpng usage in a pair of functions under PNGDriverCommon,
which convert between Godot Image and png data.
Switched to libpng 1.6 simplified API for ease of maintenance.
Implemented ImageLoaderPNG and ResourceSaverPNG in terms of
PNGDriverCommon functions.
Travis, switched to builtin libpng (thus builtin freetype and zlib also)
so we can build on Xenial.

fix for #29741
fix for #26351

ibrahn committed:
Png driver reworked to use libpng 1.6 simplified API

Wrapped libpng usage in a pair of functions under PNGDriverCommon,
which convert between Godot Image and png data.
Switched to libpng 1.6 simplified API for ease of maintenance.
Implemented ImageLoaderPNG and ResourceSaverPNG in terms of
PNGDriverCommon functions.
Travis, switched to builtin libpng (thus builtin freetype and zlib also)
so we can build on Xenial.

ibrahn commented:
Failing CI on configurations using libpng 1.2
Is 1.2 our target version at the moment?

akien-mga commented on drivers/png/image_loader_png.cpp:
```diff
@@ -29,368 +29,83 @@
 /*************************************************************************/
 
 #include "image_loader_png.h"
+#include "png_driver_common.h"
```
This should be in the next block with the core includes as per http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes.

akien-mga commented on drivers/png/png_driver_common.cpp:
```diff
@@ -0,0 +1,204 @@
+/*************************************************************************/
+
+#include "png_driver_common.h"
+#include <png.h>
```
See http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

akien-mga commented on drivers/png/png_driver_common.cpp:
```diff
@@ -0,0 +1,204 @@
+#include "core/os/os.h"
+#include <string.h>
+
+namespace PngDriverCommon {
```
I would name it `PNGDriverCommon` to match the convention used in other classes.

akien-mga commented:
> Failing CI on configurations using libpng 1.2
Is 1.2 our target version at the moment?

Travis CI builds run on Ubuntu 16.04 LTS, which still shipped with libpng12 by default (IIRC even Ubuntu 18.04 LTS ships with libpng12, only 18.10 **finally** added libpng16, half a decade after RPM-based distro...).

But we can up our minimum requirement to libpng16, which is what we bundled in `thirdparty` and what most distros provide nowadays.

The `pkg-config` check in `platform/{x11,server}/detect.py` should be changed to query `libpng16` specifically. And Travis Linux builds should use `builtin_libpng=yes`, or we should add a new repository for libpng16 on Ubuntu 16.04.

akien-mga commented:
> And Travis Linux builds should use builtin_libpng=yes, or we should add a new repository for libpng16 on Ubuntu 16.04.

The latter won't work as Travis' freetype is still linked against libpng12. So `builtin_libpng=yes` it is, and thus builtin freetype and zlib too.

akien-mga commented:
```diff
diff --git a/.travis.yml b/.travis.yml
index 587e57c74..09d8cad07 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -8,7 +8,7 @@ env:
   global:
     - SCONS_CACHE=$HOME/.scons_cache
     - SCONS_CACHE_LIMIT=1024
-    - OPTIONS="debug_symbols=no verbose=yes progress=no"
+    - OPTIONS="debug_symbols=no verbose=yes progress=no builtin_libpng=yes"
     - secure: "uch9QszCgsl1qVbuzY41P7S2hWL2IiNFV4SbAYRCdi0oJ9MIu+pVyrQdpf3+jG4rH6j4Rffl+sN17Zz4dIDDioFL1JwqyCqyCyswR8uACC0Rr8gr4Mi3+HIRbv+2s2P4cIQq41JM8FJe84k9jLEMGCGh69w+ibCWoWs74CokYVA="
 
 cache:
@@ -39,7 +39,7 @@ matrix:
             - ubuntu-toolchain-r-test
           packages:
             - &gcc8_deps [gcc-8, g++-8]
-            - &linux_deps [libasound2-dev, libfreetype6-dev, libgl1-mesa-dev, libglu1-mesa-dev, libx11-dev, libxcursor-dev, libxi-dev, libxinerama-dev, libxrandr-dev]
+            - &linux_deps [libasound2-dev, libgl1-mesa-dev, libglu1-mesa-dev, libx11-dev, libxcursor-dev, libxi-dev, libxinerama-dev, libxrandr-dev]
             - &linux_mono_deps [mono-devel, msbuild, nuget]
 
         coverity_scan:
```

ibrahn commented:
- Fixed my include formatting/order
- Renamed `PngDriverCommon` to `PNGDriverCommon`
- Changed detect.py pkg-config calls to refer to libpng16 package
- Applied travis.yml patch from @akien-mga comment above

akien-mga commented:
Thanks!
